Lilu/Headers/kern_util.hpp

```diff
@@ -944,4 +945,285 @@ class evector_base {
+	 */
+	inline bool isFull() {
+		IORecursiveLockLock(lock);
+		bool retVal = (count == size) && (indexr == indexw);
+		IORecursiveLockUnlock(lock);
+		return retVal;
+	}
+	
+	/**
+	 *  Get the number of elements in the circular buffer
+	 *
+	 *  @return The current number of elements in the buffer.
+	 */
+	inline IOItemCount getCount() {
+		IORecursiveLockLock(lock);
+		IOItemCount retVal = count;
+		IORecursiveLockUnlock(lock);
+		return retVal;
+	}
+	
+	/**
+	 *  Write the given element to the circular buffer
+	 *
+	 *  @param element The element to write
+	 *  @return `true` on success, `false` if the buffer is full.
+	 */
+	inline bool write(const T &element) {
+		if (isFull())
+			return false;
+		
+		IORecursiveLockLock(lock);
+		storage[indexw] = element;
+		indexw += 1;
+		indexw %= size;
+		count += 1;
+		IORecursiveLockUnlock(lock);
+		return true;
+	}
+	
+	/**
+	 *  Read the next element from the circular buffer
+	 *
+	 *  @param element The element read from the buffer
+	 *  @return `true` on success, `false` if the buffer is empty.
+	 */
+	inline bool read(T& element) {
+		if (isEmpty())
+			return false;
+		
+		IORecursiveLockLock(lock);
+		element = storage[indexr];
+		indexr += 1;
+		indexr %= size;
+		count -= 1;
+		IORecursiveLockUnlock(lock);
+		return true;
+	}
+};
+
+/**
+ *  Wrap an object that is not an instance of OSObject
+ */
+class OSObjectWrapper: public OSObject {
+	/**
+	 *  Constructors & Destructors
+	 */
+	OSDeclareDefaultStructors(OSObjectWrapper);
+	
+	using super = OSObject;
+	
+	/**
+	 *  Wrapped object
+	 */
+	void *object;
+	
+public:
+	/**
+	 *  Initialize the wrapper with the given object
+	 *
+	 *  @param object The wrapped object that is not an `OSObject`
+	 *  @return `true` on success, `false` otherwise.
+	 */
+	EXPORT bool init(void *object);
+	
+	/**
+	 *  Reinterpret the wrapped object as the given type
+	 *
+	 *  @return The wrapped object of the given type.
+	 */
+	template <typename T>
+	T *reinterpretWrappedObjectAs() {
+		return reinterpret_cast<T*>(object);
+	}
+	
+	/**
+	 *  Create a wrapper for the given object that is not an `OSObject`
+	 *
+	 *  @param object A non-null object
+	 *  @return A non-null wrapper on success, `nullptr` otherwise.
+	 *  @warning The caller is responsbile for managing the lifecycle of the given object.
+	 */
+	EXPORT static OSObjectWrapper *of(void *object);
+};
+
+namespace Value {
+	template <typename T>
+	struct Value {
+		T value;
+
+		explicit Value(T value) : value(value) {}
+
+	#if __cplusplus >= 201703L
+		// Available as of C++17
+		template<typename... Ts>
+		bool isOneOf(Ts... args) {
+			return ((value == args) || ...);
+		}
+
+		// Available as of C++17
+		template <typename... Ts>
+		bool isNotOneOf(Ts... args) {
+			return ((value != args) && ...);
+		}
+	#endif
+	};
+	
+	template <typename T>
+	static Value<T> of(T value) {
+		return Value<T>(value);
+	}
+}
+
+static inline uint64_t MachAbsoluteTime2Nanoseconds(uint64_t abstime) {
```

Comment:
You might be fine with https://github.com/acidanthera/Lilu/blob/master/Lilu/Headers/kern_time.hpp. If not, then better introduce convertAbsToNs and put it there.

Reply:
Yes, I am fine with the time header. I didn't notice it before.